File: src/c++/src/epp/ext/EppChangePollExt.cpp
/*******************************************************************************
 * The MIT License (MIT)
 *  
 * Copyright (c)2015 Neustar Inc.
 *  
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *  
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *  
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *******************************************************************************/

#include "EppChangePollExt.hpp"
#include "EppUtil.hpp"

EppChangePollExt::EppChangePollExt()
: state("after")
{
	reason    = null;
	operation = null;
	date      = null;
}

EppChangePollExt::~EppChangePollExt()
{
	if ( this->reason != null )
	{
		delete this->reason;
		this->reason = null;
	}
	if ( this->operation != null )
	{
		delete this->operation;
		this->operation = null;
	}
	date = null;
}

DOMElement* EppChangePollExt::toXML(DOMDocument &doc, const DOMString &tag)
{
	DOMElement* body = EppUtil::createElementNS(doc, "changePoll", "changeData");

	DOMElement* elm = null;

	if ( this->state.length() > 0 )
	{
		body->setAttribute(XS("state"), this->state);
	}

	if ( this->operation != null )
	{
		elm = this->operation->toXML(doc, "operation");
		if ( elm != null )
		{
			body->appendChild(elm);
		}
	}

	if ( this->date != null )
	{
		elm = doc.createElement(XS("date"));
		elm->appendChild(EppUtil::createTextNode(doc, date));
		body->appendChild(elm);
	}

	if ( this->svTRID.length() > 0 )
	{
		elm = doc.createElement(XS("svTRID"));
		elm->appendChild(doc.createTextNode(this->svTRID));
		body->appendChild(elm);
	}

	if ( this->who.length() > 0 )
	{
		elm = doc.createElement(XS("who"));
		elm->appendChild(doc.createTextNode(this->who));
		body->appendChild(elm);
	}

	if( this->reason != null )
	{
		if ( reason != NULL )
		{
			elm = reason->toXML(doc, "reason");
			if ( elm != null )
			{
				body->appendChild(elm);
			}
		}
	}

	return body;
}

EppChangePollExt * EppChangePollExt::fromXML( const DOMNode& root )
{
	DOMNodeList* list = root.getChildNodes();

	if ( list == NULL )
	{
		return null;
	}

	EppChangePollExt * ext = new EppChangePollExt();

	for( unsigned int i = 0; i < list->getLength(); i++ )
	{
		DOMNode* node = list->item(i);
		if ( node == NULL )
		{
			continue;
		}

		DOMString name = node->getLocalName();

		if( name.isNull() )
		{
			name = node->getNodeName();
		}

		if( name.isNull() )
		{
			continue;
		}

		if( name.equals("operation") || name.equals("changePoll:operation") )
		{
			EppChangePollOperation* oper = EppChangePollOperation::fromXML(*node);	
			ext->setOperation(oper);
		}
		else if( name.equals("date") || name.equals("changePoll:date") )
		{
			ext->setDate(EppUtil::getDate(*node));
		}
		else if( name.equals("svTRID") || name.equals("changePoll:svTRID") )
		{
			ext->setSvTRID(EppUtil::getText(*node));
		}
		else if( name.equals("who") || name.equals("changePoll:who") )
		{
			ext->setWho(EppUtil::getText(*node));
		}
		else if( name.equals("reason") || name.equals("changePoll:reason") )
		{
			EppReason* reason = EppReason::fromXML(*node);

			if ( reason != null )
			{
				ext->setReason(reason);
			}
		}
	}

	DOMNamedNodeMap* attrs = root.getAttributes();

	if ( attrs == NULL )
	{
		return ext;
	}

	for ( unsigned int i = 0; i < attrs->getLength(); i++ )
	{
		DOMNode* attr = attrs->item(i);

		if ( attr == NULL )
		{
			continue;
		}

		DOMString name = attr->getNodeName();

		if( name.equals("state") )
		{
			ext->setState(attr->getNodeValue());
			break;
		}
	}

	return ext;
}

void EppChangePollExt::setState(const DOMString& st)
{
	this->state = st;
}

DOMString EppChangePollExt::getState() const
{
	return this->state; 
}

void EppChangePollExt::setOperation(EppChangePollOperation* op)
{
	if ( this->operation != null )
	{
		delete this->operation;
	}
	this->operation = op;
}

EppChangePollOperation* EppChangePollExt::getOperation() const
{
	return this->operation;
}

void EppChangePollExt::setDate(time_t dt)
{
	this->date = dt;
}

time_t EppChangePollExt::getDate() const
{
	return this->date;
}

void EppChangePollExt::setSvTRID(const DOMString& trid)
{
	this->svTRID = trid;
}

DOMString EppChangePollExt::getSvTRID() const
{
	return this->svTRID;
}

void EppChangePollExt::setWho(const DOMString& wh)
{
	this->who = wh;
}

DOMString EppChangePollExt::getWho() const
{
	return this->who; 
}

void EppChangePollExt::setReason(EppReason* reas)
{
	if ( this->reason != null )
	{ 
		delete this->reason;
	}
	this->reason = reas;
}

EppReason* EppChangePollExt::getReason() const
{
	return this->reason;
}

File: src/java/src/com/neulevel/epp/ext/EppChangePollExt.java
/*******************************************************************************
 * The MIT License (MIT)
 *  
 * Copyright (c) 2015 Neustar Inc.
 *  
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *  
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *  
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *******************************************************************************/
package com.neulevel.epp.ext;

import java.util.Calendar;

import org.w3c.dom.Attr;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import com.neulevel.epp.core.EppUtil;
import com.neulevel.epp.core.response.EppReason;

/**
 * 
 * @author ankit.kothana
 *
 */
public class EppChangePollExt extends EppExtension {
	
	private String state;
	private EppChangePollOperation eppChangePollOperation;
	private Calendar date;
	private String who;
	private String serverTransId;
	private EppReason eppReason;
	

	public EppChangePollExt() {
	}

	public EppChangePollExt(String state,
			EppChangePollOperation eppChangePollOperation, Calendar date,
			String who, String serverTransId, EppReason eppReason) {
		this.state = state;
		this.eppChangePollOperation = eppChangePollOperation;
		this.date = date;
		this.who = who;
		this.serverTransId = serverTransId;
		this.eppReason = eppReason;
	}

	/**
	 * @return the state
	 */
	public String getState() {
		return state;
	}

	/**
	 * @param state the state to set
	 */
	public void setState(String state) {
		this.state = state;
	}

	/**
	 * @return the eppChangePollOperation
	 */
	public EppChangePollOperation getEppChangePollOperation() {
		return eppChangePollOperation;
	}

	/**
	 * @param eppChangePollOperation the eppChangePollOperation to set
	 */
	public void setEppChangePollOperation(
			EppChangePollOperation eppChangePollOperation) {
		this.eppChangePollOperation = eppChangePollOperation;
	}

	/**
	 * @return the date
	 */
	public Calendar getDate() {
		return date;
	}

	/**
	 * @param date the date to set
	 */
	public void setDate(Calendar date) {
		this.date = date;
	}

	/**
	 * @return the who
	 */
	public String getWho() {
		return who;
	}

	/**
	 * @param who the who to set
	 */
	public void setWho(String who) {
		this.who = who;
	}

	/**
	 * @return the serverTransId
	 */
	public String getServerTransId() {
		return serverTransId;
	}

	/**
	 * @param serverTransId the serverTransId to set
	 */
	public void setServerTransId(String serverTransId) {
		this.serverTransId = serverTransId;
	}

	/**
	 * @return the eppReason
	 */
	public EppReason getEppReason() {
		return eppReason;
	}

	/**
	 * @param eppReason the eppReason to set
	 */
	public void setEppReason(EppReason eppReason) {
		this.eppReason = eppReason;
	}

	@Override
	public Element toXML(Document doc, String tag) {
		Element body = EppUtil.createElementNS(doc, "changePoll", "changeData");
		Element elem = null;
		if (state != null && state.length() > 0) {
			Attr attr = doc.createAttribute("state");
			attr.setValue( this.state );
			body.setAttributeNode(attr); 
		}
		
		if (eppChangePollOperation != null) {
			elem = eppChangePollOperation.toXML(doc, "operation");
			if (elem != null) {
				body.appendChild(elem);
			}
		}
		
		if (date != null) {
			elem = doc.createElement("date");
			elem.appendChild(EppUtil.createTextNode(doc, date));
			body.appendChild(elem);
		}
		
		if (serverTransId != null) {
			elem = doc.createElement("svTRID");
			elem.appendChild(doc.createTextNode(serverTransId));
			body.appendChild(elem);
		}
		
		if (who != null) {
			elem = doc.createElement("who");
			elem.appendChild(doc.createTextNode(who));
			body.appendChild(elem);
		}
		
		if (eppReason != null) {
			elem = eppReason.toXML(doc, "reason");
			if (elem != null) {
				body.appendChild(elem);
			}
		}
		
		return body;
	}
	
	public static EppChangePollExt fromXML(final Node root) {
		EppChangePollExt returnEppChangePollExt = new EppChangePollExt();
		NodeList nodeList = root.getChildNodes();
		for (int i = 0; i < nodeList.getLength(); i++) {
			Node node = nodeList.item(i);
			String name = node.getLocalName();
			if (name == null)
			{
				name = node.getNodeName();
			}

			if (name != null)
			{
				if (name.equals("operation"))
				{
					EppChangePollOperation changePollOp = EppChangePollOperation.fromXML(node);
					returnEppChangePollExt.setEppChangePollOperation(changePollOp);
					continue;
				}

				if (name.equals("date"))
				{
					returnEppChangePollExt.setDate(EppUtil.getDate(node));
					continue;
				}
				
				if (name.equals("svTRID"))
				{
					returnEppChangePollExt.setServerTransId(EppUtil.getText(node));
					continue;
				}
				
				if (name.equals("who"))
				{
					returnEppChangePollExt.setWho(EppUtil.getText(node));
					continue;
				}
				
				if (name.equals("reason"))
				{
					EppReason reason = EppReason.fromXML(node);
					returnEppChangePollExt.setEppReason(reason);
					continue;
				}
			} else {
				continue;
			}
		}
		
		NamedNodeMap attrs = root.getAttributes();
		for(int i = 0;i<attrs.getLength();i++) {
			Node attr = attrs.item(i);
			String attrValue = attr.getNodeValue();
			if( attr.getNodeName().equals("state") ) {
				returnEppChangePollExt.setState(attrValue);
				continue;
			} 
			
		}
		
		return returnEppChangePollExt;
	}

	/* (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	@Override
	public String toString() {
		return new String("changePoll");
	}


}
